File: textimageview/src/main/java/com/antonionicolaspina/textimageview/TextImageView.java
package com.antonionicolaspina.textimageview;

import android.annotation.TargetApi;
import android.content.Context;
import android.content.res.Resources;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.graphics.PointF;
import android.graphics.Rect;
import android.graphics.RectF;
import android.graphics.Typeface;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.ScaleGestureDetector;
import android.widget.ImageView;

import java.util.ArrayList;

public class TextImageView extends ImageView implements ScaleGestureDetector.OnScaleGestureListener, RotationGestureDetector.OnRotationGestureListener {
  public interface OnTextMovedListener {
    void textMoved(PointF position);
  }
  protected static class TextProperties {
    private String text;

    public float scale;
    public float size;
    public String[] textLines;
    public Paint paint;
    public ArrayList<Rect> textRects;
    public PointF textPosition;
    public PointF rotationCenter;
    public float rotation;

    public TextProperties(String text, float size, int color) {
      this.scale = 1f;
      this.size = size;
      this.paint = new Paint(Paint.ANTI_ALIAS_FLAG);
      this.textRects = new ArrayList<>();
      this.textPosition = new PointF(0f, 0f);
      this.rotationCenter = new PointF();
      this.rotation = 0f;

      paint.setColor(color);
      paint.setTextSize(size);
      setText(text);
    }

    public void setText(String text) {
      this.text = text;
      this.textLines = null;
      if (null != text) {
        this.textLines = text.split("\n");

        textRects.clear();
        for(int i=0; i<textLines.length; i++) {
          Rect r = new Rect();
          paint.getTextBounds(textLines[i], 0, textLines[i].length(), r);
          textRects.add(i, r);
        }
      }
    }

    public String getText() {
      return text;
    }
  }

  public enum ClampMode {UNLIMITED, ORIGIN_INSIDE, TEXT_INSIDE}

  private ScaleGestureDetector scaleDetector;
  private RotationGestureDetector rotateDetector;

  // region Global parameters
  private float minSize;
  private float maxSize;
  private boolean panEnabled;
  private boolean scaleEnabled;
  private boolean rotationEnabled;
  private ClampMode clampTextMode;
  private int interline;
  private RectF imageRect;
  // endregion

  // region Other members
  private PointF focalPoint;
  private OnTextMovedListener onTextMovedListener;
  private float previousRotation = 0f;
  private ArrayList<TextProperties> texts;
  private int currentSize;
  private int currentColor;
  // endregion

  public TextImageView(Context context) {
    super(context);
    init(context, null);
  }

  public TextImageView(Context context, AttributeSet attrs) {
    super(context, attrs);
    init(context, attrs);
  }

  public TextImageView(Context context, AttributeSet attrs, int defStyleAttr) {
    super(context, attrs, defStyleAttr);
    init(context, attrs);
  }

  @TargetApi(Build.VERSION_CODES.LOLLIPOP)
  public TextImageView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
    super(context, attrs, defStyleAttr, defStyleRes);
    init(context, attrs);
  }

  protected void init(Context context, AttributeSet attributeSet) {
    texts = new ArrayList<>();

    imageRect  = new RectF();
    focalPoint = new PointF();

    Resources resources = context.getResources();

    if (null != attributeSet) {
      TypedArray attrs    = context.getTheme().obtainStyledAttributes(attributeSet, R.styleable.TextImageView, 0, 0);
      currentSize = attrs.getDimensionPixelSize(R.styleable.TextImageView_android_textSize, resources.getDimensionPixelSize(R.dimen.default_text_size));
      currentColor = attrs.getColor(R.styleable.TextImageView_android_textColor, Color.BLACK);
      panEnabled = attrs.getBoolean(R.styleable.TextImageView_tiv_panEnabled, false);
      scaleEnabled = attrs.getBoolean(R.styleable.TextImageView_tiv_scaleEnabled, false);
      rotationEnabled = attrs.getBoolean(R.styleable.TextImageView_tiv_rotationEnabled, false);
      interline = attrs.getDimensionPixelOffset(R.styleable.TextImageView_tiv_interline, 0);
      clampTextMode = ClampMode.values()[attrs.getInt(R.styleable.TextImageView_tiv_clampTextMode, 0)];
      setText(attrs.getString(R.styleable.TextImageView_android_text));

      minSize = attrs.getDimensionPixelSize(R.styleable.TextImageView_tiv_minTextSize, resources.getDimensionPixelSize(R.dimen.default_min_text_size));
      maxSize = attrs.getDimensionPixelSize(R.styleable.TextImageView_tiv_maxTextSize, resources.getDimensionPixelSize(R.dimen.default_max_text_size));
      attrs.recycle();
    }

    scaleDetector  = new ScaleGestureDetector(context, this);
    rotateDetector = new RotationGestureDetector(this);
  }

  @Override
  protected void onDraw(Canvas canvas) {
    super.onDraw(canvas);

    if ( isInEditMode() && (0 == texts.size()) ) {
      setText("sample text");
    }

    // Get rectangle of the drawable
    imageRect.top  = 0;
    imageRect.left = 0;

    Drawable drawable = getDrawable();
    if (null != drawable) {
      imageRect.right = drawable.getIntrinsicWidth();
      imageRect.bottom = drawable.getIntrinsicHeight();
    }
    // Translate and scale the rectangle
    getImageMatrix().mapRect(imageRect);

    for(TextProperties tp: texts) {
      canvas.save();
      if (rotationEnabled) {
        canvas.rotate(-tp.rotation, tp.rotationCenter.x, tp.rotationCenter.y);
      }

      // Draw text
      float top = tp.textPosition.y + imageRect.top;
      for (int i = 0; i < tp.textLines.length; i++) {
        int h = tp.textRects.get(i).height();
        canvas.save();
        canvas.translate(tp.textPosition.x + imageRect.left, top + h);
        canvas.drawText(tp.textLines[i], 0, 0, tp.paint);
        canvas.restore();
        top += h + interline * tp.scale;
      }
      canvas.restore();
    }
  }

  protected void recalculateFocalPoint(MotionEvent event) {
    final int pointerCount = event.getPointerCount();
    if (pointerCount <= 0) {
      return;
    }

    focalPoint.x = 0f;
    focalPoint.y = 0f;
    for(int i=0; i<pointerCount; i++) {
      focalPoint.x += event.getX(i);
      focalPoint.y += event.getY(i);
    }
    focalPoint.x /= pointerCount;
    focalPoint.y /= pointerCount;
  }

  protected static float between(float value, float min, float max) {
    return Math.max(Math.min(value, max), min);
  }

  @Override
  public boolean onTouchEvent(MotionEvent event) {
    scaleDetector.onTouchEvent(event);
    rotateDetector.onTouchEvent(event);
    super.onTouchEvent(event);

    final int action = event.getAction();
    switch(action & MotionEvent.ACTION_MASK) {
      case MotionEvent.ACTION_DOWN:
      case MotionEvent.ACTION_UP:
      case MotionEvent.ACTION_POINTER_UP:
        recalculateFocalPoint(event);
        return true;
      case MotionEvent.ACTION_MOVE:
        if (0 < texts.size()) {
          final float x = focalPoint.x;
          final float y = focalPoint.y;
          TextProperties tp = texts.get(texts.size() - 1);

          recalculateFocalPoint(event);

          if (panEnabled) {
            tp.textPosition.x += focalPoint.x - x;
            tp.textPosition.y += focalPoint.y - y;

            tp.rotationCenter.x += focalPoint.x - x;
            tp.rotationCenter.y += focalPoint.y - y;

            reclampText();

            invalidate();
          }
        }
        return true;
    }
    return false;
  }

  protected void reclampText() {
    if (0 == texts.size()) {
      return;
    }

    TextProperties tp = texts.get(texts.size()-1);
    switch (clampTextMode) {
      case UNLIMITED:
        break;
      case ORIGIN_INSIDE: {
        RectF enclosingRect = calculateEnclosingRect();
        enclosingRect.offset(-imageRect.left, -imageRect.top);
        tp.textPosition.x -= enclosingRect.left-between(enclosingRect.left, 0, imageRect.width());
        tp.textPosition.y -= enclosingRect.top-between(enclosingRect.top, 0, imageRect.height());
        invalidate();
        break;
      }
      case TEXT_INSIDE: {
        RectF enclosingRect = calculateEnclosingRect();
        enclosingRect.offset(-imageRect.left, -imageRect.top);
        tp.textPosition.x -= enclosingRect.left - between(enclosingRect.left, 0, imageRect.width()-enclosingRect.width());
        tp.textPosition.y -= enclosingRect.top - between(enclosingRect.top, 0, imageRect.height()-enclosingRect.height());
        invalidate();
        break;
      }
    }

    if (null != onTextMovedListener) {
      PointF position = getTextPosition();
      if ( (!Float.isNaN(position.x)) && (!Float.isNaN(position.y)) ) {
        onTextMovedListener.textMoved(position);
      }
    }
  }

  protected RectF calculateEnclosingRect() {
    if (0 == texts.size()) {
      return null;
    }

    TextProperties tp = texts.get(texts.size()-1);

    Matrix mat = new Matrix();
    RectF globalRect = new RectF();
    float top = tp.textPosition.y;
    for(int i=0; i<tp.textLines.length; i++) {
      int h = tp.textRects.get(i).height();
      RectF rect = new RectF(0, 0, tp.textRects.get(i).width(), h);
      rect.offset(imageRect.left, imageRect.top);

      mat.reset();
      mat.preRotate(-tp.rotation, tp.rotationCenter.x, tp.rotationCenter.y);
      mat.preTranslate(tp.textPosition.x, top);

      mat.mapRect(rect);

      if (0 == i) {
        globalRect.set(rect);
      } else {
        globalRect.top = Math.min(globalRect.top, rect.top);
        globalRect.left = Math.min(globalRect.left, rect.left);
        globalRect.bottom = Math.max(globalRect.bottom, rect.bottom);
        globalRect.right = Math.max(globalRect.right, rect.right);
      }
      top += h + interline*tp.scale;
    }

    return globalRect;
  }

  @Override
  public boolean onScale(ScaleGestureDetector scaleGestureDetector) {
    if (scaleEnabled && (0<texts.size())) {
      TextProperties tp = texts.get(texts.size()-1);
      tp.scale *= scaleGestureDetector.getScaleFactor();
      tp.paint.setTextSize(Math.max(minSize, Math.min(tp.scale * tp.size, maxSize)));
      tp.scale = tp.paint.getTextSize() / tp.size;
      tp.setText(tp.text);
      reclampText();
      invalidate();
    }

    return true;
  }

  @Override
  public boolean onScaleBegin(ScaleGestureDetector scaleGestureDetector) {
    return true;
  }

  @Override
  public void onScaleEnd(ScaleGestureDetector scaleGestureDetector) {
  }

  @Override
  public void OnRotation(RotationGestureDetector rotationDetector) {
    if (rotationEnabled && (0<texts.size())) {
      TextProperties tp = texts.get(texts.size()-1);
      tp.rotation += rotationDetector.getAngle() - previousRotation;
      previousRotation = rotationDetector.getAngle();

      tp.rotationCenter.x = focalPoint.x;
      tp.rotationCenter.y = focalPoint.y;
      invalidate();
    }
  }

  /**************
   *** Public ***
   **************/

  /**
   * Set text to be drawn over the image.
   * @param text The text.
   */
  public void setText(String text) {
    texts.clear();
    addText(text);
  }

  /**
   * Adds a text to be drawn over the image, above existing texts.
   * @param text The text.
   */
  public void addText(String text) {
    if (null != text) {
      texts.add(new TextProperties(text, currentSize, currentColor));
      reclampText();
      invalidate();
    }
  }

  /**
   * Removes the text on the top of the stack.
   */
  public void removeText() {
    if (0 < texts.size()) {
      texts.remove(texts.size()-1);
      invalidate();
    }
  }

  /**
   * Set the typeface to use for the text.
   * @param typeface The typeface to be used.
   */
  public void setTypeface(Typeface typeface) {
    TextProperties tp = texts.get(texts.size()-1);
    tp.paint.setTypeface(typeface);
    reclampText();
    invalidate();
  }

  /**
   * Enable or disable user-pan for the view.
   * @param enabled Whether panning should be enabled.
   */
  public void setPanEnabled(boolean enabled) {
    panEnabled = enabled;
  }

  /**
   * Enable or disable user-scaling for the view.
   * @param enabled Whether scaling should be enabled.
   */
  public void setScaleEnabled(boolean enabled) {
    scaleEnabled = enabled;
  }

  /**
   * Enable or disable user-rotation for the view.
   * @param enabled Whether rotation should be enabled.
   */
  public void setRotationEnabled(boolean enabled) {
    rotationEnabled = enabled;
  }

  /**
   * Set the text color.
   * @param color Color in the format of <a href="http://developer.android.com/reference/android/graphics/Color.html">android.graphics.Color</a>.
   *
   * @see <a href="http://developer.android.com/reference/android/graphics/Color.html">android.graphics.Color</a>
   */
  public void setTextColor(int color) {
    TextProperties tp = texts.get(texts.size()-1);
    tp.paint.setColor(color);
    invalidate();
  }

  /**
   * Set the default text size to the given value, interpreted as "scaled pixel" units.
   * This size is adjusted based on the current density and user font size preference.
   * @param textSize The scaled pixel size.
   */
  public void setTextSize(float textSize) {
    TextProperties tp = texts.get(texts.size()-1);
    tp.scale = 1f;
    tp.size  = textSize;
    tp.paint.setTextSize(textSize);
    tp.setText(tp.text);
    reclampText();
    invalidate();
  }

  /**
   * Return offset position between the text and the image. Considers both top left corners to the the calculation.
   * @return Pointf containing x and y offsets, as a per-one value. Eg. (0,0)=top-left, (1,1)=bottom-right.
   */
  public PointF getTextPosition() {
    RectF enclosingRect = calculateEnclosingRect();
    enclosingRect.offset(-imageRect.left, -imageRect.top);
    return new PointF(enclosingRect.left / imageRect.width(), enclosingRect.top / imageRect.height());
  }

  /**
   * Set the listener to be fired when the text changes its location.
   * @param listener the listener to be called, or null.
   */
  public void setOnTextMovedListener(OnTextMovedListener listener) {
    this.onTextMovedListener = listener;
  }

  /**
   * Get the relative size between the image and the text.
   * @return Relative size. Eg. 0.5=text half the height of the image.
   */
  public float getTextRelativeSize() {
    TextProperties tp = texts.get(texts.size()-1);
    return tp.paint.getTextSize() / imageRect.height();
  }
}

File: textimageview/src/main/java/com/antonionicolaspina/textimageview/TextImageView.kt
package com.antonionicolaspina.textimageview

import android.animation.ValueAnimator
import android.annotation.SuppressLint
import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.animation.LinearInterpolator
import androidx.annotation.ColorInt
import androidx.appcompat.widget.AppCompatImageView
import androidx.core.content.ContextCompat
import androidx.core.graphics.contains
import androidx.core.graphics.toPoint
import androidx.core.graphics.toRectF
import androidx.core.graphics.withMatrix
import com.mapbox.android.gestures.*
import kotlin.math.*

class TextImageView
@JvmOverloads constructor(
  context: Context, attributeSet: AttributeSet? = null, defStyleAttr: Int = 0
): AppCompatImageView(context, attributeSet, defStyleAttr) {
  interface Listener {
    fun textsChanged(texts: List<Text>)
    fun textTapped(text: Text)
  }

  var panEnabled = false
  var scaleEnabled = false
  var rotationEnabled = false
  val deleteEnabled: Boolean

  var listener: Listener? = null
  var initialTextSize = 0f
  var minTextSize = 0f
  var maxTextSize = 0f

  private val texts = mutableListOf<TextProperties>()
  private var selectedText: TextProperties? = null

  private var deleteAreaVisible = false
  private val deleteAreaHeight = context.resources.getDimensionPixelSize(R.dimen.tiv_delete_area_height)
  private val deleteButtonHeight = context.resources.getDimensionPixelSize(R.dimen.tiv_delete_button_height)
  private val deleteDrawable = ContextCompat.getDrawable(context, R.drawable.ic_trash)!!
  private var deleteButtonScale = 1f

  init {
    context.theme.obtainStyledAttributes(attributeSet, R.styleable.TextImageView, 0, 0).apply {
      try {
        panEnabled = getBoolean(R.styleable.TextImageView_tiv_panEnabled, false)
        scaleEnabled = getBoolean(R.styleable.TextImageView_tiv_scaleEnabled, false)
        rotationEnabled = getBoolean(R.styleable.TextImageView_tiv_rotationEnabled, false)
        deleteEnabled = getBoolean(R.styleable.TextImageView_tiv_deleteEnabled, false)

        initialTextSize = getDimensionPixelSize(R.styleable.TextImageView_tiv_initialTextSize, resources.getDimensionPixelSize(
            R.dimen.tiv_default_text_size
          )).toFloat()
        minTextSize = getDimensionPixelSize(R.styleable.TextImageView_tiv_minTextSize, resources.getDimensionPixelSize(
            R.dimen.tiv_default_min_text_size
          )).toFloat()
        maxTextSize = getDimensionPixelSize(R.styleable.TextImageView_tiv_maxTextSize, resources.getDimensionPixelSize(R.dimen.tiv_default_max_text_size)).toFloat()
      } finally {
        recycle()
      }
    }
    if (deleteEnabled) {
      setPadding(
        paddingLeft,
        paddingTop,
        paddingRight,
        paddingBottom+deleteAreaHeight
      )
    }
  }

  override fun onDraw(canvas: Canvas) {
    super.onDraw(canvas)

    if (isInEditMode) {
      if (texts.isEmpty()) {
        setText("sample text")
      }
      deleteAreaVisible = deleteEnabled
    }

    if (deleteEnabled && deleteAreaVisible) {
      val w = measuredWidth
      val h = measuredHeight

      val buttonSize = getDeleteButtonSize()
      deleteDrawable.setBounds(
        (w-buttonSize)/2,
        h-deleteAreaHeight+(deleteAreaHeight-buttonSize)/2,
        (w+buttonSize)/2,
        h-(deleteAreaHeight-buttonSize)/2
      )
      deleteDrawable.draw(canvas)
    }

    texts.forEach { tp ->
      canvas.withMatrix(tp.matrix) {
        canvas.drawText(tp.text, 0f, 0f, tp.paint)
      }
    }
  }

  /**
   * Set text to be drawn over the image.
   * @param text The text.
   */
  fun setText(text: String) {
    texts.clear()
    addText(text)
  }

  /**
   * Change selected text.
   * @param text The text.
   */
  fun changeText(text: String) {
    selectedText?.let {
      it.text = text
      textChanged(it)
    }
  }

  /**
   * Set the typeface to use for the text.
   * @param typeface The typeface to be used.
   */
  fun setTypeface(typeface: Typeface?) {
    selectedText?.let {
      it.paint.typeface = typeface
      textChanged(it)
    }
  }

  fun setPosition(position: PointF) {
    selectedText?.let {
      it.position.x = position.x*measuredWidth
      it.position.y = position.y*measuredHeight
      textChanged(it)
    }
  }

  fun setScaleFactor(scale: Float) {
    selectedText?.let {
      it.scaleFactor = scale
      textChanged(it)
    }
  }

  fun setRotationDegrees(rotation: Float) {
    selectedText?.let {
      it.rotationDegress = rotation
      textChanged(it)
    }
  }

  fun getSelected() = selectedText?.toText(measuredWidth, measuredHeight)

  /**
   * Set the text color.
   * @param color Color in the format of <a href="http://developer.android.com/reference/android/graphics/Color.html">android.graphics.Color</a>.
   *
   * @see <a href="http://developer.android.com/reference/android/graphics/Color.html">android.graphics.Color</a>
   */
  fun setTextColor(@ColorInt color: Int) {
    selectedText?.let {
      it.paint.color = color
      textChanged(it)
    }
  }

  /**
   * Adds a text to be drawn over the image, above existing texts.
   * @param text The text.
   */
  fun addText(text: String) {
    val tp = TextProperties(text)
    tp.paint.textSize = initialTextSize
    tp.paint.getTextBounds(tp.text, 0, tp.text.length, tp.boundingRect)
    tp.position.x = measuredWidth/2f-tp.boundingRect.exactCenterX()
    tp.position.y = measuredHeight/2f-tp.boundingRect.exactCenterY()

    texts.add(tp)
    selectedText = tp
    textChanged(tp)
  }

  /**
   * Adds a drop shadow
   */
  fun addDropShadow(
    angle: Float = 45f, distance: Float = 0.02f, @ColorInt color: Int = Color.argb(
      0x80,
      0,
      0,
      0
    )
  ) {
    selectedText?.let {
      val textSize = it.paint.textSize
      it.paint.setShadowLayer(
        1f,
        textSize * cos(angle) * distance,
        textSize * sin(angle) * distance,
        color
      )
    }
  }

  /**
   * Returns all texts' data
   */
  fun getTexts(): List<Text> {
    val w = measuredWidth
    val h = measuredHeight
    return texts.map { it.toText(w, h) }
  }

  private var valueAnimator: ValueAnimator? = null
  //region Gestures
  private val androidGesturesManager = AndroidGesturesManager(context).apply {
    setMoveGestureListener(object : MoveGestureDetector.SimpleOnMoveGestureListener() {
      override fun onMove(
        detector: MoveGestureDetector,
        distanceX: Float,
        distanceY: Float
      ): Boolean {
        if (panEnabled) {
          selectedText?.let { tp ->
            tp.position.x -= distanceX
            tp.position.y -= distanceY
            textChanged(tp)

            if (deleteEnabled && deleteAreaVisible) {
              if (inDeleteArea(tp)) {
                ValueAnimator.ofFloat(deleteButtonScale, 1.5f)
                  .setDuration(200)
                  .apply {
                    interpolator = LinearInterpolator()
                    addUpdateListener {
                      deleteButtonScale = it.animatedValue as Float
                      invalidate()
                    }
                    start()
                    valueAnimator = this
                  }
              } else {
                ValueAnimator.ofFloat(deleteButtonScale, 1f)
                  .setDuration(200)
                  .apply {
                    interpolator = LinearInterpolator()
                    addUpdateListener {
                      deleteButtonScale = it.animatedValue as Float
                      invalidate()
                    }
                    start()
                    valueAnimator = this
                  }
              }
            }
          }
        }
        return true
      }

      override fun onMoveBegin(detector: MoveGestureDetector): Boolean {
        if (deleteEnabled) {
          if (null != selectedText) {
            deleteButtonScale = 1f
            deleteAreaVisible = true
            invalidate()
          }
        }
        return true
      }

      override fun onMoveEnd(detector: MoveGestureDetector, velocityX: Float, velocityY: Float) {
        if (deleteEnabled && deleteAreaVisible) {
          deleteAreaVisible = false
          if (true == valueAnimator?.isRunning) {
            valueAnimator?.cancel()
            valueAnimator = null
          }
          deleteButtonScale = 1f
          selectedText?.let {
            checkDeleted(it)
          }
          invalidate()
        }
      }
    })

    setStandardScaleGestureListener(object :
      StandardScaleGestureDetector.SimpleStandardOnScaleGestureListener() {
      override fun onScale(detector: StandardScaleGestureDetector): Boolean {
        if (scaleEnabled) {
          selectedText?.let {
            val s = it.scaleFactor * detector.scaleFactor
            val textSize = s * it.paint.textSize
            it.scaleFactor = max(minTextSize, min(textSize, maxTextSize)) / it.paint.textSize
            textChanged(it)
          }
        }
        return true
      }
    })

    setRotateGestureListener(object : RotateGestureDetector.SimpleOnRotateGestureListener() {
      override fun onRotate(
        detector: RotateGestureDetector,
        rotationDegreesSinceLast: Float,
        rotationDegreesSinceFirst: Float
      ): Boolean {
        if (rotationEnabled) {
          selectedText?.let {
            it.rotationDegress -= rotationDegreesSinceLast
            textChanged(it)
          }
        }
        return true
      }
    })

    setStandardGestureListener(object : StandardGestureDetector.SimpleStandardOnGestureListener() {
      override fun onDown(e: MotionEvent): Boolean {
        texts.forEach { tp ->
          val p = tp.inverseMap(e.x, e.y)

          if (tp.boundingRect.contains(p.toPoint())) {
            selectedText = tp
          }
        }
        invalidate()
        return true
      }

      override fun onSingleTapConfirmed(e: MotionEvent): Boolean {
        val listener = listener
        val selectedText = selectedText
        if ( (null!=selectedText) && (null != listener) ) {
          val p = selectedText.inverseMap(e.x, e.y)

          if (selectedText.boundingRect.contains(p.toPoint())) {
            listener.textTapped(selectedText.toText(measuredWidth, measuredHeight))
          }
        }
        return true
      }
    })
  }

  private fun getDeleteButtonSize() = (deleteButtonHeight*deleteButtonScale).roundToInt()
  private fun inDeleteArea(tp: TextProperties): Boolean {
    val w = measuredWidth
    val h = measuredHeight
    val buttonSize = getDeleteButtonSize()
    val r = RectF(
      (w-buttonSize)/2f,
      h-deleteAreaHeight+(deleteAreaHeight-buttonSize)/2f,
      (w+buttonSize)/2f,
      h-(deleteAreaHeight-buttonSize)/2f
    )

    val boundingRect = tp.boundingRect.toRectF()
    tp.matrix.mapRect(boundingRect)
    return r.intersect(boundingRect)
  }

  private fun checkDeleted(tp: TextProperties) {
    if (inDeleteArea(tp)) {
      texts.remove(tp)
      invalidate()
      listener?.textsChanged(getTexts())
    }
  }

  @SuppressLint("ClickableViewAccessibility")
  override fun onTouchEvent(event: MotionEvent?): Boolean {
    androidGesturesManager.onTouchEvent(event)
    return true
  }
  //endregion

  private fun textChanged(tp: TextProperties) {
    invalidate()

    tp.paint.getTextBounds(tp.text, 0, tp.text.length, tp.boundingRect)

    with(tp.matrix) {
      reset()
      setScale(tp.scaleFactor, tp.scaleFactor)
      preTranslate(tp.position.x / tp.scaleFactor, tp.position.y / tp.scaleFactor)
      preRotate(
        tp.rotationDegress,
        tp.boundingRect.exactCenterX() / tp.scaleFactor,
        tp.boundingRect.exactCenterY() / tp.scaleFactor
      )
    }

    listener?.textsChanged(getTexts())
  }
}
